keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1037 @@
+
+For Pods carrying `.spec.schedulingPaused=true`, they will be "parked" in scheduler's internal
+unschedulablePods pool, and only get tried when the field is set to `false`.
+
+Practically, this field can be initialized to `true` by a MutatingWebhook, and be flipped to `false`
+by external integrators when certain criteria is met.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an orchestrator developer, such as a dynamic quota manager, I have the full picutre to know when
+Pods are scheduling-ready; therefore, I want an API to signal to kube-scheduler when to consider a
+Pod for scheduling. The pattern for this story would be to use a mutating webhook to force creating
+pods in a "not-ready to schedule" state that the custom orchestrator changes to ready at a later
+time based on its own evaluations.
+
+#### Story 2
+
+This story is an extension of the previous one in that more than one custom orchestrator could be
+deployed on the same cluster, therefore they want an API that enables establishing an agreement om
+when a pod is considered ready for scheduling.
+
+#### Story 3
+
+As an advanced scheduler developer, I want to compose a series of scheduler PreEnqueue plugins
+to guard the schedulability of my Pods. This enables splitting custom enqueue admission logic into
+several building blocks, and thus offers the most flexibility. Meanwhile, some plugin needs to visit
+the in-memory state (like waiting Pods) that is only accessible via scheduler framework.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- **Restricted state transition:** The `schedulingPaused` state can only transit to `true` upon
+Pod's creation, and can transit to `false` afterwards once.
+
+    | schedulingPaused=true⏸️ | schedulingPaused=false▶️ |
+    |------------------------|-------------------------|
+    | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
+
+- **v1 guarantee:** If the feature is enabled and the new field is updated to `true`, it's still
+supported to set the pod's .spec.nodeName without clearing the `schedulingPaused` flag. However,
+scheduled Pods that carry `schedulingPaused=true` is a warning signal revealing one of the
+following reasons:
+    - The feature is not enabled in default scheduler
+    - Default scheduler is at a lower version than API Server
+    - A custom scheduler or administrator may have enforce setting nodeName for the Pod
+Operation-wise, you need to report this behavior to your cluster administrator.
```

Comment:
Will any component report a metric when this situation is observed?

Reply:
Is it permitted to set `schedulingPaused` to `true` _and_ set `.spec.nodeName` at create time? Would be good to make the situation clear.

Reply:
> schedulingPaused to true and set .spec.nodeName at create time

we should reject that because the user intent is contradictory.

but we should allow unpausing and setting the nodeName in the same update request. With `schedulingGates`,  this means an update that results in no gates remaining can also include setting the nodeName.

Reply:
@sftim we had a discussion on "whether or not to allow setting nodeName for a paused Pod" back and forth. The final conclusion is no (the original version). The wording in this version is out-of-dated, sry for the confusion. I'm going to update it today.

Reply:
(In terms of my feedback here, I think this is resolved).